src/node_dns.cc

```diff
@@ -0,0 +1,471 @@
+           Allocator* a = reinterpret_cast<Allocator*>(data);
+           a->free(ptr, a->data);
+         });
+  GETDNS(extension_set_libuv_loop, isolate, context_, env->event_loop());
+  GETDNS(context_set_tls_authentication,
+         isolate,
+         context_,
+         GETDNS_AUTHENTICATION_REQUIRED);
+}
+
+DNSWrap::~DNSWrap() {
+  getdns_context_destroy(context_);
+  CHECK_EQ(current_getdns_memory_, 0);
+}
+
+void DNSWrap::MemoryInfo(MemoryTracker* tracker) const {
+  // tracker->TrackFieldWithSize("getdns_memory", current_getdns_memory_, "getdns_context");
+}
+
+void DNSWrap::CheckAllocatedSize(size_t previous_size) const {
+  CHECK_GE(current_getdns_memory_, previous_size);
+}
+
+void DNSWrap::IncreaseAllocatedSize(size_t size) {
+  current_getdns_memory_ += size;
+}
+
+void DNSWrap::DecreaseAllocatedSize(size_t size) {
+  current_getdns_memory_ -= size;
+}
+
+void DNSWrap::New(const FunctionCallbackInfo<Value>& args) {
+  CHECK(args.IsConstructCall());
+  Environment* env = Environment::GetCurrent(args);
+  new DNSWrap(env, args.This());
+}
+
+// static
+void DNSWrap::Callback(getdns_context* getdns_context,
+                       getdns_callback_type_t cb_type,
+                       getdns_dict* response,
+                       void* data,
+                       getdns_transaction_t tid) {
+  DNSWrap* dns = reinterpret_cast<DNSWrap*>(data);
+  Isolate* isolate = dns->env()->isolate();
+  HandleScope scope(isolate);
+  Local<Context> context = dns->env()->context();
+  Context::Scope context_scope(context);
+
+  auto cleanup = OnScopeLeave([&]{
+    getdns_dict_destroy(response);
+  });
+
+  switch (cb_type) {
+    case GETDNS_CALLBACK_COMPLETE:
+      break;
+#define V(name)                                                                \
+  case GETDNS_CALLBACK_##name:                                                 \
+    dns->RejectTransaction(                                                    \
+        tid, ERR_DNS_##name(isolate, GETDNS_CALLBACK_##name##_TEXT));          \
+    return;
+      V(CANCEL)
+      V(TIMEOUT)
+      V(ERROR)
+#undef V
+    default:
+      UNREACHABLE();
+  }
+
+  {
+    uint32_t status;
+    GETDNS(dict_get_int, isolate, response, "/status", &status);
+    switch (status) {
+      case GETDNS_RESPSTATUS_GOOD:
+        break;
+#define V(name)                                                                \
+  case GETDNS_RESPSTATUS_##name:                                               \
+    dns->RejectTransaction(                                                    \
+        tid, ERR_DNS_##name(isolate, GETDNS_RESPSTATUS_##name##_TEXT));        \
+    return;
+        V(NO_NAME)
+        V(ALL_TIMEOUT)
+        V(NO_SECURE_ANSWERS)
+        V(ALL_BOGUS_ANSWERS)
+#undef V
+      default:
+        UNREACHABLE();
+    }
+  }
+
+  char* json = getdns_print_json_dict(response, 0);
+  Local<String> result = String::NewFromUtf8(isolate, json).ToLocalChecked();
+  free(json);
```

Comment:
Given that this is semver-major already, I'm curious if instead of doing the JSON conversion here... can we wrap the `response` in an object and just return that, allowing the user code to decide when to extract the details? Or, is that just too much additional complexity here?